[gui][fonts] font code cleanup

AlwinEsch opened the pull request:
## Description

Related to https://github.com/xbmc/xbmc/pull/20552 and before it comes in, is here a cleanup about font system.

This here does not change the workflow about fonts and should not bring problems about changes.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[gui][fonts] replace all NULL uses to C++ nullptr

Here in font files, all "C" associated NULLs are changed to "C ++" nullptr.

AlwinEsch committed:
[gui][fonts] cleanup log calls

Previously, in some cases, only a single line was made without a reference to the function, which makes it harder to recognize what it belongs to.

AlwinEsch committed:
[gui][fonts] rename value strFileName to fontIdent

There becomes a formatted string given with `StringUtils::Format("{}_{:f}_{:f}{}", strFilename, newSize, aspect, fontInfo.border ? "_border" : "");` where not relates to an filename direct!
As them formatted is it used to identify it on future places.

Here becomes it renamed to prevent this perverse big confusing part, as there in e.g. GUIFontTTF one time defined "m_strFilename" (contains true path) and further also "m_strFileName" where contains a id.

Seems also that there already was come issues and bugs about, as there the normal font becomes asked on places with formated text.
About this need bigger investigate whats there.

CastagnaIT commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -327,17 +332,19 @@ CGUIFont* GUIFontManager::GetDefaultFont(bool border)
-      font13border = LoadTTF("__defaultborder__", fontInfo.fileName, 0xFF000000, 0, fontInfo.size, font13->GetStyle(), true, 1.0f, fontInfo.aspect, &fontInfo.sourceRes, fontInfo.preserveAspect);
+      font13border = LoadTTF("__defaultborder__", fontInfo.fileName, 0xFF000000, 0, fontInfo.size,
+                             font13->GetStyle(), true, 1.0f, fontInfo.aspect, &fontInfo.sourceRes,
+                             fontInfo.preserveAspect);
```
`0xFF000000` can be replaced by `UTILS::COLOR::BLACK`

AlwinEsch replied:
Thanks for the note, has changed to `UTILS::COLOR::BLACK`.

CastagnaIT commented on xbmc/guilib/GUIFont.cpp:
```diff
@@ -99,23 +103,25 @@ void CGUIFont::DrawText(float x,
   {
     shadowColor = CServiceBroker::GetWinSystem()->GetGfxContext().MergeAlpha(shadowColor);
     std::vector<UTILS::COLOR::Color> shadowColors;
     for (unsigned int i = 0; i < renderColors.size(); i++)
```
unsigned int -> size_t

CastagnaIT commented on xbmc/guilib/GUIFont.cpp:
```diff
@@ -193,32 +202,36 @@ void CGUIFont::DrawScrollingText(float x,
     offset = scrollInfo.m_totalWidth - scrollInfo.pixelPos;
 
   std::vector<UTILS::COLOR::Color> renderColors;
   for (unsigned int i = 0; i < colors.size(); i++)
```
unsigned int -> size_t

CastagnaIT commented on xbmc/guilib/GUIFont.cpp:
```diff
@@ -193,32 +202,36 @@ void CGUIFont::DrawScrollingText(float x,
   {
     shadowColor = CServiceBroker::GetWinSystem()->GetGfxContext().MergeAlpha(shadowColor);
     std::vector<UTILS::COLOR::Color> shadowColors;
     for (unsigned int i = 0; i < renderColors.size(); i++)
```
unsigned int -> size_t

CastagnaIT commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -650,11 +611,9 @@ float CGUIFontTTF::GetLineHeight(float lineSpacing) const
+  return m_cellHeight + SPACING_BETWEEN_CHARACTERS_IN_TEXTURE;
 }
 
 std::vector<CGUIFontTTF::Glyph> CGUIFontTTF::GetHarfBuzzShapedGlyphs(const vecText& text)
```
i cannot comment below this line but there are others loops `unsigned int` should be `size_t`

AlwinEsch replied:
Has changed it with additional commit, only as it uses negative values (-1) and `hb_buffer_add` as `unsigned int` has changed one from `int` to `unsigned int` and about the size it use now a `static_cast`.

CastagnaIT reviewed (commented):
only minor things, i leave the word to the most experienced

thexai commented on xbmc/guilib/GUIFontTTFDX.cpp:
```diff
@@ -176,13 +179,15 @@ void CGUIFontTTFDX::LastEnd()
   CD3DBuffer* buffer = nullptr;
-  if (!vertices.empty()) // do not create empty buffers, leave buffer as nullptr, it will be ignored on drawing stage
+  if (!vertices
+           .empty()) // do not create empty buffers, leave buffer as nullptr, it will be ignored on drawing stage
```
Maybe move comment to upper line to avoid this rare line break

thexai commented on xbmc/guilib/GUIFontTTFDX.h:
```diff
@@ -47,20 +47,23 @@ class CGUIFontTTFDX : public CGUIFontTTF, public ID3DResource
-  std::unique_ptr<CD3DTexture> m_speedupTexture; // extra texture to speed up reallocations
-  Microsoft::WRL::ComPtr<ID3D11Buffer> m_vertexBuffer;
+  unsigned m_vertexWidth{0};
+  std::unique_ptr<CD3DTexture> m_speedupTexture{nullptr}; // extra texture to speed up reallocations
```
`std::unique_ptr` no need to initialize with nullptr, it's already done by default

thexai commented on xbmc/guilib/GUIFontTTFDX.h:
```diff
@@ -47,20 +47,23 @@ class CGUIFontTTFDX : public CGUIFontTTF, public ID3DResource
-  Microsoft::WRL::ComPtr<ID3D11Buffer> m_vertexBuffer;
+  unsigned m_vertexWidth{0};
+  std::unique_ptr<CD3DTexture> m_speedupTexture{nullptr}; // extra texture to speed up reallocations
+  Microsoft::WRL::ComPtr<ID3D11Buffer> m_vertexBuffer{nullptr};
```
`Microsoft::WRL::ComPtr` no need to initialize with nullptr, it's already done by default

thexai commented on xbmc/guilib/GUIFontCache.cpp:
```diff
@@ -241,6 +241,6 @@ template void CGUIFontCache<CGUIFontCacheDynamicPosition, CGUIFontCacheDynamicVa
 void CVertexBuffer::clear()
 {
-  if (m_font != NULL)
+  if (m_font != nullptr)
```
Maybe `if (m_font)`

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -341,7 +341,7 @@ bool CGUIFontTTF::Load(
 void CGUIFontTTF::Begin()
 {
-  if (m_nestedBeginCount == 0 && m_texture != NULL && FirstBegin())
+  if (m_nestedBeginCount == 0 && m_texture != nullptr && FirstBegin())
```
Maybe 

`&& m_texture &&`

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -896,7 +896,7 @@ bool CGUIFontTTF::CacheCharacter(wchar_t letter, uint32_t style, Character* ch,
 
         std::unique_ptr<CTexture> newTexture = ReallocTexture(newHeight);
-        if (newTexture == NULL)
+        if (newTexture == nullptr)
```
`if (!newTexture)`

thexai commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -907,7 +907,7 @@ bool CGUIFontTTF::CacheCharacter(wchar_t letter, uint32_t style, Character* ch,
     }
 
-    if (m_texture == NULL)
+    if (m_texture == nullptr)
```
`if (!m_texture)`

thexai commented on xbmc/guilib/GUIFontTTFGL.cpp:
```diff
@@ -346,12 +346,12 @@ std::unique_ptr<CTexture> CGUIFontTTFGL::ReallocTexture(unsigned int& newHeight)
       CTexture::CreateTexture(m_textureWidth, newHeight, XB_FMT_A8);
 
-  if (!newTexture || newTexture->GetPixels() == NULL)
+  if (!newTexture || newTexture->GetPixels() == nullptr)
```
`|| !newTexture->GetPixels())`

AlwinEsch committed:
[gui][fonts] cast on "for" loop by CGUIFontTTF::GetHarfBuzzShapedGlyphs

There was in "for" loops "int" and "unsigned int" used and where amount gives "size_t" as value.
This adds a "static_cast" and leaved by "int" and "unsigned int" as one use negative value and "hb_buffer_add" used the "unsigned int" as value.

AlwinEsch committed:
[gui][fonts] make construction value set in header

This shifts the value setting of the variables from the constructions function to their headers, making it cleaner.

AlwinEsch reviewed (commented):
Thanks @CastagnaIT and @thexai to look about!

Have updated the commits and also added two additional new commits.

AlwinEsch committed:
[gui][fonts] not use nullptr on "if" check

Due it only by value itself.

AlwinEsch committed:
[gui][fonts] not store m_strFilename in CGUIFontTTF

This path value was nowhere used and then is memory used where never needed.

garbear reviewed (approved):
I prefer being explicit with nullptr to use pointer arithmetic instead of boolean logic for a pointer

fuzzard commented on xbmc/guilib/GUIFont.cpp:
```diff
@@ -98,24 +102,26 @@ void CGUIFont::DrawText(float x,
-  if (!shadowColor) shadowColor = m_shadowColor;
+  for (const auto& color : colors)
+    renderColors.emplace_back(
+        CServiceBroker::GetWinSystem()->GetGfxContext().MergeAlpha(color ? color : m_textColor));
```
GetWinSystem is a pointer, therefore has a possibility to be null. Really should check for null imo

AlwinEsch replied:
To make for here it becomes to heavy, by the really big amount of `CServiceBroker::GetWinSystem()` calls must be there a another way, instead of a `if (null)`.

fuzzard replied:
essentially you cant chain the ->function into the getwinsystem, as it'll just crash if a null ptr is returned by the getwinsystem (ie the call is then nullptr->GetGfcContext() = crash).

Also keep in mind, this was all brought about due to a headless future possibly. if we have no win system, why would we want to blindly run gui related stuff, therefore if getwinsystem == nullptr, we move on without processing, hence the checks

AlwinEsch replied:
@fuzzard have added on last commit a pointer check about, can you look over it?

fuzzard commented on xbmc/guilib/GUIFontCache.h:
```diff
@@ -121,16 +122,12 @@ struct CGUIFontCacheHash
-           a.m_scrolling == b.m_scrolling &&
-           a.m_maxPixelWidth == b.m_maxPixelWidth &&
+    return a.m_text == b.m_text && a.m_colors == b.m_colors && a.m_alignment == b.m_alignment &&
+           a.m_scrolling == b.m_scrolling && a.m_maxPixelWidth == b.m_maxPixelWidth &&
            Match(a.m_pos, a.m_matrix, b.m_pos, b.m_matrix, a.m_scrolling) &&
-           a.m_scaleX == b.m_scaleX &&
-           a.m_scaleY == b.m_scaleY;
+           a.m_scaleX == b.m_scaleX && a.m_scaleY == b.m_scaleY;
```
id argue this makes it harder to read.

AlwinEsch replied:
Added a clang ignore for here and below.

fuzzard commented on xbmc/guilib/GUIFontCache.h:
```diff
@@ -243,30 +243,30 @@ struct CVertexBuffer
+         diffY - floorf(diffY) < 2 * FONT_CACHE_DIST_LIMIT &&
+         diffZ - floorf(diffZ) < 2 * FONT_CACHE_DIST_LIMIT && a_m.m[0][0] == b_m.m[0][0] &&
+         a_m.m[1][1] == b_m.m[1][1] && a_m.m[2][2] == b_m.m[2][2];
+  // We already know the first 3 columns of both matrices are diagonal, so no need to check the other elements
```
same again here. Id lean more towards the original layout

fuzzard commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -158,20 +157,25 @@ CGUIFont* GUIFontManager::LoadTTF(const std::string& strFontName,
 
   // font file is loaded, create our CGUIFont
-  CGUIFont *pNewFont = new CGUIFont(strFontName, iStyle, textColor, shadowColor, lineSpacing, (float)iSize, pFontFile);
+  CGUIFont* pNewFont = new CGUIFont(strFontName, iStyle, textColor, shadowColor, lineSpacing,
+                                    (float)iSize, pFontFile);
   m_vecFonts.push_back(pNewFont);
```
Just a comment, but maybe an opportunity to change from a new/delete call into a vector of smart pointers. at a glance the new/delete stuff all line up, so probably not absolutely required.

AlwinEsch replied:
Have added on new commit an change to use `shared_ptr` about.

On tests it was OK, only found a memleak about `CGUIButtonControl` by change of language within settings.

One class is not removed and then remains in Kodi as a memleak. Could narrow it down in search to seems likely to come from CGUIDialogSettingsBase (presumably here https://github.com/xbmc/xbmc/blob/master/xbmc/settings/dialogs/GUIDialogSettingsBase.cpp#L371, but not sure?)

It would probably be good to revise the controls in another request in order to also have `shared_mem` there, with the huge number of GUIControls it would be really useful and safer.

CastagnaIT commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -117,7 +115,8 @@ CGUIFont* GUIFontManager::LoadTTF(const std::string& strFontName,
   std::string strPath;
```
above ^^^line 111 use C++ cast

CastagnaIT commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -222,10 +227,10 @@ bool GUIFontManager::OnMessage(CGUIMessage &message)
+  for (size_t i = 0; i < m_vecFonts.size(); ++i)
   {
     CGUIFont* font = m_vecFonts[i];
     OrigFontInfo fontInfo = m_vecFontInfo[i];
```
line 239 C++ cast

AlwinEsch replied:
Have added the missing cast's.

lrusak commented on xbmc/guilib/GUIFont.cpp:
```diff
@@ -170,8 +177,10 @@ void CGUIFont::DrawScrollingText(float x,
-  if (!shadowColor) shadowColor = m_shadowColor;
+  if (!m_font)
+    return;
+  if (!shadowColor)
```
My only comment on this PR would be that I would like to see some extra whitespace to help make it easier to read. In places like this it would be nice to have a newline after the return. There are quite a few places that could be updated but I'm only going to comment about this once 👍 Of course this is just my opinion and doesn't impact the functionality at all.

AlwinEsch replied:
Would be better, have added some new lines on places like this.

AlwinEsch committed:
[gui][fonts] add some new code line spaces

To makes it more easy in overview and better as to have all short together.

AlwinEsch commented:
Should be done by now.

The only shame is that there is a memleak with CButtonControl, but would not fit in here and I hope that this will be attacked soon. Thank god only when changing the language.

Is someone already changing the GUI controls to shared_mem?

AlwinEsch committed:
[gui][fonts] remove dead unused function CGUIFont::RemapGlyph

AlwinEsch committed:
[gui][fonts] confirm CServiceBroker::GetWinSystem() != nullptr

This changes the calls to "CServiceBroker::GetWinSystem()" and checks if they give a nullptr.

In addition, in order to reduce the number of calls to "CServiceBroker::GetWinSystem()->GetGfxContext ()", the `CGraphicContext` class is stored separately and thus greatly reduces the number of CPU calls.

Furthermore, the underlying function calls which use `CGraphicContext` are passed this class as a function value in order to avoid a nullptr check at the called point.

lrusak commented on xbmc/guilib/GUIFont.cpp:
```diff
@@ -17,19 +17,21 @@
-    for (vecText::size_type i = 0; i < wsuffix.size(); i++)
-      suffix.push_back(wsuffix[i]);
-    Reset();
+  initialWait = wait;
```
It would also be nice to initialize them after the constructor
```
 : m_initialPos(pos),
...
```

AlwinEsch replied:
Has changed and now done in constructor.

lrusak commented on xbmc/guilib/GUIFont.cpp:
```diff
@@ -17,19 +17,21 @@
+  g_charsetConverter.utf8ToW(scrollSuffix, wsuffix);
+  suffix.clear();
+  suffix.reserve(wsuffix.size());
+  for (vecText::size_type i = 0; i < wsuffix.size(); i++)
```
This stuff can probably be replaced with something like:
```
m_suffix = vecText(wsuffix.begin(), wsuffix.end());
```

AlwinEsch replied:
Have added it and seems to work.

lrusak commented on xbmc/guilib/GUIFont.h:
```diff
@@ -34,34 +36,34 @@ class CGUIFontTTF;
-  {
-    pixelSpeed = speed * 0.001f;
-  }
+  void SetSpeed(int speed) { pixelSpeed = speed * 0.001f; }
```
```suggestion
  void SetSpeed(int speed) { m_pixelSpeed = speed * 0.001f; }
```

lrusak commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -142,37 +153,41 @@ CGUIFont* GUIFontManager::LoadTTF(const std::string& strFontName,
-                  strFontName, strFilename);
-        return LoadTTF(strFontName, "arial.ttf", textColor, shadowColor, iSize, iStyle, border, lineSpacing, originalAspect);
+        CLog::Log(
+            LOGERROR,
```
This formatting seems odd.

AlwinEsch replied:
There was done by the clang cleanup and no direct idea to make shorter.

Maybe as:
```cpp
        CLog::Log(LOGERROR, "GUIFontManager::{}: Couldn't load font name: {}({}), trying arial.ttf",
                  __func__, strFontName, strFilename);
```
with the reduced length bring clang it as this.

AlwinEsch replied:
Have added the reduced length. If not OK then I remove again.

lrusak commented on xbmc/guilib/GUIFontManager.cpp:
```diff
@@ -142,37 +153,41 @@ CGUIFont* GUIFontManager::LoadTTF(const std::string& strFontName,
   // font file is loaded, create our CGUIFont
-  CGUIFont *pNewFont = new CGUIFont(strFontName, iStyle, textColor, shadowColor, lineSpacing, (float)iSize, pFontFile);
-  m_vecFonts.push_back(pNewFont);
+  std::shared_ptr<CGUIFont> pNewFont =
```
good candidate for `auto`

AlwinEsch replied:
Makes it shorter, good idea and clang is happy.

lrusak commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -152,38 +143,19 @@ class CFreeTypeLibrary
-  m_nTexture = 0;
-
-  m_renderSystem = CServiceBroker::GetRenderSystem();
+  m_vertex.reserve(4 * 1024);
```
magic number? any ideas why we allocate `4 * 1024`?

AlwinEsch replied:
Really magic, no idea why it was set as this.

lrusak commented on xbmc/guilib/GUIFontTTF.h:
```diff
@@ -68,29 +67,45 @@ struct SVertex
+class CGUIFontTTF : public std::enable_shared_from_this<CGUIFontTTF>
 {
+  static constexpr size_t LOOKUPTABLE_SIZE = 256 * 8;
+  static constexpr int CHARS_PER_TEXTURE_LINE = 20; // number characters to cache per texture line
```
These should go into the cpp file in an anonymous namespace and they don't need to be static

lrusak commented on xbmc/guilib/GUIFontTTF.h:
```diff
@@ -119,8 +134,8 @@ class CGUIFontTTF
   {
-    int startOffset;
-    int endOffset;
+    unsigned int startOffset;
```
```suggestion
    unsigned int m_startOffset;
```

AlwinEsch replied:
Has done on some of the the different structure values, hope Windows becomes happy.

lrusak commented on xbmc/guilib/GUIFontTTFDX.h:
```diff
@@ -22,7 +22,7 @@
 #include <wrl/client.h>
 
-#define ELEMENT_ARRAY_MAX_CHAR_INDEX (2000)
+constexpr size_t ELEMENT_ARRAY_MAX_CHAR_INDEX = 2000;
```
move to anonymous namespace in cpp file

lrusak commented on xbmc/guilib/GUIFontTTFGL.cpp:
```diff
@@ -82,7 +78,7 @@ bool CGUIFontTTFGL::FirstBegin()
   {
     // Have OpenGL generate a texture object handle for us
-    glGenTextures(1, (GLuint*) &m_nTexture);
+    glGenTextures(1, (GLuint*)&m_nTexture);
```
```suggestion
    glGenTextures(1, static_cast<GLuint*>(&m_nTexture));
```

lrusak commented on xbmc/guilib/GUIFontTTFGL.cpp:
```diff
@@ -193,25 +197,28 @@ void CGUIFontTTFGL::LastEnd()
-    glVertexAttribPointer(colLoc,  4, GL_UNSIGNED_BYTE, GL_TRUE, sizeof(SVertex), (char*)vertices + offsetof(SVertex, r));
-    glVertexAttribPointer(tex0Loc, 2, GL_FLOAT,  GL_FALSE, sizeof(SVertex), (char*)vertices + offsetof(SVertex, u));
+    glVertexAttribPointer(posLoc, 3, GL_FLOAT, GL_FALSE, sizeof(SVertex),
+                          (char*)vertices + offsetof(SVertex, x));
```
```suggestion
                          static_cast<char*>(vertices) + offsetof(SVertex, x));
```

lrusak reviewed (commented):
I've added some comments about some minors I saw. If they are too much for this PR maybe a second follow up would be good.

I like the use of `shared_from_this()`, however, we need to be careful that all references are deleted. It would be nice to get rid of the manual reference counting. I say this because I tried briefly to use `std::shared_ptr<CGUIFont>` and had some troubles. Hopefully you make it work 👍

lrusak commented on xbmc/guilib/GUIFontTTFGL.h:
```diff
@@ -15,30 +15,36 @@
 
 #include "system_gl.h"
 
+constexpr size_t ELEMENT_ARRAY_MAX_CHAR_INDEX = 1000;
```
anonymous namespace in the cpp file

scott967 commented:
When I tried to compile this on Windows got an error \xbmc\guilib\GUIFontTTFDX.cpp(195): error C2440: '<function-style-cast>':
 cannot convert from 'initializer list' to 'CVertexBuffer'

AlwinEsch committed:
[gui][fonts] replace #define use to constexpr

Here the "C" associated "#define" path is replaced by the C++ associated `constexpr` path.

AlwinEsch committed:
[gui][fonts] add on some structure values a "m_" to begin

Also few other cleanups included.

AlwinEsch commented:
> When I tried to compile this on Windows got an error \xbmc\guilib\GUIFontTTFDX.cpp(195): error C2440: '': cannot convert from 'initializer list' to 'CVertexBuffer'

Have here no Windows available at the moment and was come in as error, after I become Jenkins working would be really good if you test it on Windows and thanks a lot that you look about :+1:.

AlwinEsch committed:
[gui][fonts] make value re set and use to use static_cast and auto

In some places the old "C" was used away to change from e.g. float to interger, here it is changed to use static_cast for it.
In addition, in a few for loops it has been changed to use "auto", or in some which "i" need it, it has been changed to "size_t", which meant that a few "C" casts could be removed.

AlwinEsch commented:
After reinstall of Windows by me, has done a test and was running on Windows by this changes.

AlwinEsch committed:
[gui][fonts] change vector push_back to emplace_back

AlwinEsch committed:
[gui][fonts] complete clang cleanup about font files

Here all source code files belonging to the font are cleaned up using clang and so that they are correct in the current style. Nothing has changed in the processing path.

AlwinEsch committed:
[gui][fonts] store used fonts in std::unique_ptr

This secures the list management and in some circumstances avoids a mem leak if the list is emptied without delete in the entry.

The move to "GUIFontManager::GUIFontManager() = default;" was needed by this error on test:
```
In file included from /usr/include/c++/11/memory:76,
                 from /home/alwin/Development/Kodi/kodi-N/xbmc/utils/logtypes.h:11,
                 from /home/alwin/Development/Kodi/kodi-N/xbmc/settings/lib/ISettingControl.h:11,
                 from /home/alwin/Development/Kodi/kodi-N/xbmc/settings/SettingControl.h:11,
                 from /home/alwin/Development/Kodi/kodi-N/xbmc/settings/Settings.h:12,
                 from /home/alwin/Development/Kodi/kodi-N/xbmc/settings/Settings.cpp:9:
/usr/include/c++/11/bits/unique_ptr.h: In instantiation of ‘void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = CGUIFont]’:
/usr/include/c++/11/bits/unique_ptr.h:361:17:   required from ‘std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = CGUIFont; _Dp = std::default_delete<CGUIFont>]’
/usr/include/c++/11/bits/stl_construct.h:140:22:   required from ‘constexpr void std::_Destroy(_Tp*) [with _Tp = std::unique_ptr<CGUIFont>]’
/usr/include/c++/11/bits/stl_construct.h:152:19:   required from ‘static void std::_Destroy_aux<<anonymous> >::__destroy(_ForwardIterator, _ForwardIterator) [with _ForwardIterator = std::unique_ptr<CGUIFont>*; bool <anonymous> = false]’
/usr/include/c++/11/bits/stl_construct.h:185:11:   required from ‘void std::_Destroy(_ForwardIterator, _ForwardIterator) [with _ForwardIterator = std::unique_ptr<CGUIFont>*]’
/usr/include/c++/11/bits/alloc_traits.h:746:15:   required from ‘void std::_Destroy(_ForwardIterator, _ForwardIterator, std::allocator<_T2>&) [with _ForwardIterator = std::unique_ptr<CGUIFont>*; _Tp = std::unique_ptr<CGUIFont>]’
/usr/include/c++/11/bits/stl_vector.h:680:15:   required from ‘std::vector<_Tp, _Alloc>::~vector() [with _Tp = std::unique_ptr<CGUIFont>; _Alloc = std::allocator<std::unique_ptr<CGUIFont> >]’
/home/alwin/Development/Kodi/kodi-N/xbmc/guilib/GUIFontManager.h:50:3:   required from ‘static std::shared_ptr<_Tp> xbmcutil::GlobalsSingleton<T>::getInstance() [with T = GUIFontManager]’
/home/alwin/Development/Kodi/kodi-N/xbmc/guilib/GUIFontManager.h:106:1:   required from here
/usr/include/c++/11/bits/unique_ptr.h:83:23: error: invalid application of ‘sizeof’ to incomplete type ‘CGUIFont’
   83 |         static_assert(sizeof(_Tp)>0,
      |                       ^~~~~~~~~~~
/usr/include/c++/11/bits/unique_ptr.h: In instantiation of ‘void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = CGUIFontTTF]’:
/usr/include/c++/11/bits/unique_ptr.h:361:17:   required from ‘std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = CGUIFontTTF; _Dp = std::default_delete<CGUIFontTTF>]’
/usr/include/c++/11/bits/stl_construct.h:140:22:   required from ‘constexpr void std::_Destroy(_Tp*) [with _Tp = std::unique_ptr<CGUIFontTTF>]’
/usr/include/c++/11/bits/stl_construct.h:152:19:   required from ‘static void std::_Destroy_aux<<anonymous> >::__destroy(_ForwardIterator, _ForwardIterator) [with _ForwardIterator = std::unique_ptr<CGUIFontTTF>*; bool <anonymous> = false]’
/usr/include/c++/11/bits/stl_construct.h:185:11:   required from ‘void std::_Destroy(_ForwardIterator, _ForwardIterator) [with _ForwardIterator = std::unique_ptr<CGUIFontTTF>*]’
/usr/include/c++/11/bits/alloc_traits.h:746:15:   required from ‘void std::_Destroy(_ForwardIterator, _ForwardIterator, std::allocator<_T2>&) [with _ForwardIterator = std::unique_ptr<CGUIFontTTF>*; _Tp = std::unique_ptr<CGUIFontTTF>]’
/usr/include/c++/11/bits/stl_vector.h:680:15:   required from ‘std::vector<_Tp, _Alloc>::~vector() [with _Tp = std::unique_ptr<CGUIFontTTF>; _Alloc = std::allocator<std::unique_ptr<CGUIFontTTF> >]’
/home/alwin/Development/Kodi/kodi-N/xbmc/guilib/GUIFontManager.h:50:3:   required from ‘static std::shared_ptr<_Tp> xbmcutil::GlobalsSingleton<T>::getInstance() [with T = GUIFontManager]’
/home/alwin/Development/Kodi/kodi-N/xbmc/guilib/GUIFontManager.h:106:1:   required from here
/usr/include/c++/11/bits/unique_ptr.h:83:23: error: invalid application of ‘sizeof’ to incomplete type ‘CGUIFontTTF’
```

AlwinEsch commented:
@lrusak have changed to use unique_ptr about used fonts.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Acceptable to bring in? So that I can finish the other request.

lrusak reviewed (approved):
Great work! I've had a look over this a few times and this PR is a large improvement to the code. The author has been great with review comments. This PR has been through a few rounds of review and I trust the author has verified that the code still works.

👍
